library/x509_create.c

```diff
@@ -125,81 +125,288 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
     return( cur );
 }
 
-int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
+static int x509_dn_check_special_chars( const char **cur, const char *end )
 {
     int ret = 0;
-    const char *s = name, *c = s;
-    const char *end = s + strlen( s );
-    const char *oid = NULL;
-    const x509_attr_descriptor_t* attr_descr = NULL;
-    int in_tag = 1;
+
+    /* Check for valid escaped characters */
+    if( *cur == end || *( *cur ) != ',' )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+
+    return( ret );
+}
+
+static int x509_dn_parse_tag( const char **cur, const char *end,
+                              const x509_attr_descriptor_t **attr_descr )
+{
+    int ret = 0;
+    const char *start = *cur;
+
+    /*
+     * As defined by RFC 1779, the DN is constructed from
+     * <string> | <key> <optional-space> "=" <optional-space> <string>
+     * where optional space is
+     * ( <CR> ) *( " " )
+     *
+     * When parsing the tag:
+     * 1. Skip whitespaces.
+     * 2. Read the characters, to evaluate as a valid tag.
+     * 3. Skip whitespaces until reaching "=" character.
+     */
+
+    /*
+     * Skip whitespaces
+     */
+    while( *cur < end && *( *cur ) == ' ' )
+    {
+        ( *cur )++;
+        start++;
+    }
+
+    /*
+     * Read the next characters to evaluate as a valid tag.
+     */
+    while( *cur < end && *( *cur ) != '=' && *( *cur ) !=' ' )
+    {
+         ( *cur )++;
+    }
+
+    /*
+     * cur character is either '=' or ' '
+     * or we reached the end.
+     */
+
+    if( *cur == end )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME ;
+        goto exit;
+    }
+
+    if( ( *attr_descr = x509_attr_descr_from_name
+                        ( start, *cur - start ) ) == NULL )
```

Comment:
Minor suggestion: In this cases where the assignment and conditional do not fit in the same line, I would suggest simply doing the assignment in a separate statement to help readability. For example:

```
*attr_descr = x509_attr_descr_from_name( start, *cur - start );
if(  *attr_descr == NULL )
```